File: scripts/upload-and-mint.ts
#!/usr/bin/env ts-node
/**
 * upload-and-mint.ts
 *
 * One-command flow that:
 *   1. Computes SHA-256 of the model weights file
 *   2. Uploads weights to IPFS via NFT.Storage
 *   3. (Optionally) uploads weights to Arweave via Irys
 *   4. Builds the ERC721AI metadata JSON
 *   5. Uploads metadata JSON to IPFS
 *   6. Calls mintModel() on the deployed ERC721AI contract
 *
 * Usage:
 *   npx ts-node scripts/upload-and-mint.ts \
 *     --weights model.bin \
 *     --architecture "ResNet-50" \
 *     --dataset-hash <sha256hex> \
 *     [--name "My Model"] \
 *     [--description "Fine-tuned on ImageNet"] \
 *     [--arweave]              # enable Arweave upload
 *     [--rpc-url <url>]        # override RPC
 *     [--contract <address>]   # override contract address
 */

import { createReadStream, readFileSync } from "fs";
import { createHash } from "crypto";
import { program } from "commander";
import { ethers } from "ethers";
import "dotenv/config";

// ── Types ───────────────────────────────────────────────────────────────────

interface UploadResult {
  ipfsCid: string;
  arweaveTxn: string;
  metadataCid: string;
  modelHash: string;
}

// ── Helpers ─────────────────────────────────────────────────────────────────

/** Compute SHA-256 of a file (returns hex string). */
function sha256File(filePath: string): string {
  const data = readFileSync(filePath);
  return createHash("sha256").update(data).digest("hex");
}

/** Upload a Buffer to IPFS via the NFT.Storage HTTP API and return the CID. */
async function uploadToIPFS(
  data: Buffer,
  apiKey: string,
  filename: string
): Promise<string> {
  // NFT.Storage upload endpoint
  const res = await fetch("https://api.nft.storage/upload", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/octet-stream",
      "X-Filename": filename,
    },
    body: data,
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`NFT.Storage upload failed (${res.status}): ${text}`);
  }

  const json: any = await res.json();
  return json.value.cid as string;
}

/** Upload a Buffer to Arweave via Irys and return the transaction ID. */
async function uploadToArweave(
  data: Buffer,
  irysRpcUrl: string,
  privateKey: string
): Promise<string> {
  // Dynamic import so the dependency is only needed when --arweave is used.
  const { default: Irys } = await import("@irys/sdk");
  const irys = new Irys({
    url: irysRpcUrl,
    token: "ethereum",
    key: privateKey,
  });

  await irys.ready();

  const price = await irys.getPrice(data.length);
  console.log(`  Arweave upload cost: ${irys.utils.fromAtomic(price)} ETH`);

  const receipt = await irys.upload(data, {
    tags: [{ name: "Content-Type", value: "application/octet-stream" }],
  });

  return receipt.id;
}

// ── CLI ─────────────────────────────────────────────────────────────────────

program
  .requiredOption("--weights <path>", "Path to model weights file")
  .requiredOption("--architecture <arch>", "Model architecture description")
  .requiredOption("--dataset-hash <hash>", "SHA-256 hex of training dataset")
  .option("--name <name>", "Human-readable model name", "AI Model")
  .option("--description <desc>", "Model description", "")
  .option("--arweave", "Also upload weights to Arweave via Irys", false)
  .option("--rpc-url <url>", "Ethereum JSON-RPC URL")
  .option("--contract <address>", "Deployed ERC721AI contract address")
  .option("--dry-run", "Skip on-chain mint (useful for testing uploads)", false);

async function main() {
  program.parse();
  const opts = program.opts();

  const weightsPath: string = opts.weights;
  const architecture: string = opts.architecture;
  const datasetHash: string = opts.datasetHash;
  const modelName: string = opts.name;
  const description: string = opts.description;
  const useArweave: boolean = opts.arweave;
  const dryRun: boolean = opts.dryRun;

  const rpcUrl =
    opts.rpcUrl || process.env.SEPOLIA_RPC_URL || "http://127.0.0.1:8545";
  const contractAddress =
    opts.contract || process.env.ERC721AI_CONTRACT_ADDRESS;
  const nftStorageKey = process.env.NFT_STORAGE_API_KEY;
  const irysRpcUrl =
    process.env.IRYS_RPC_URL || "https://node1.irys.xyz";
  const deployerKey = process.env.DEPLOYER_PRIVATE_KEY;

  if (!nftStorageKey) throw new Error("NFT_STORAGE_API_KEY not set");
  if (!dryRun && !contractAddress)
    throw new Error("ERC721AI_CONTRACT_ADDRESS not set");
  if (!dryRun && !deployerKey)
    throw new Error("DEPLOYER_PRIVATE_KEY not set");

  // 1. Hash the weights file
  console.log("⏳ Computing SHA-256 of weights file…");
  const modelHash = sha256File(weightsPath);
  console.log(`  Model hash: ${modelHash}`);

  // 2. Upload weights to IPFS
  console.log("⏳ Uploading weights to IPFS…");
  const weightsData = readFileSync(weightsPath);
  const ipfsCid = await uploadToIPFS(weightsData, nftStorageKey, "model.bin");
  console.log(`  IPFS CID: ${ipfsCid}`);

  // 3. (Optional) upload to Arweave
  let arweaveTxn = "";
  if (useArweave) {
    if (!process.env.IRYS_PRIVATE_KEY)
      throw new Error("IRYS_PRIVATE_KEY not set");
    console.log("⏳ Uploading weights to Arweave via Irys…");
    arweaveTxn = await uploadToArweave(
      weightsData,
      irysRpcUrl,
      process.env.IRYS_PRIVATE_KEY
    );
    console.log(`  Arweave txn: ${arweaveTxn}`);
  }

  // 4. Build metadata JSON
  const metadata = {
    name: modelName,
    description,
    model_hash: modelHash,
    storage_cid: ipfsCid,
    arweave_txn: arweaveTxn,
    architecture,
    training_dataset_hash: datasetHash,
  };
  console.log("⏳ Uploading metadata JSON to IPFS…");
  const metadataBuf = Buffer.from(JSON.stringify(metadata, null, 2));
  const metadataCid = await uploadToIPFS(
    metadataBuf,
    nftStorageKey,
    "metadata.json"
  );
  const metadataURI = `ipfs://${metadataCid}`;
  console.log(`  Metadata URI: ${metadataURI}`);

  if (dryRun) {
    console.log("\n🏁 Dry-run complete. Results:");
    console.log(JSON.stringify({ modelHash, ipfsCid, arweaveTxn, metadataURI }, null, 2));
    return;
  }

  // 5. Mint on-chain
  console.log("⏳ Minting token on-chain…");
  const provider = new ethers.JsonRpcProvider(rpcUrl);
  const signer = new ethers.Wallet(deployerKey!, provider);

  const abi = [
    "function mintModel(address to, bytes32 modelHash, string storageCid, string arweaveTxn, string architecture, bytes32 datasetHash, string metadataURI) external returns (uint256)",
  ];
  const contract = new ethers.Contract(contractAddress!, abi, signer);

  const modelHashBytes = "0x" + modelHash;
  const datasetHashBytes = "0x" + datasetHash;

  const tx = await contract.mintModel(
    signer.address,
    modelHashBytes,
    ipfsCid,
    arweaveTxn,
    architecture,
    datasetHashBytes,
    metadataURI
  );
  const receipt = await tx.wait();

  console.log(`\n✅ Minted! tx: ${receipt.hash}`);
  console.log(
    JSON.stringify(
      { modelHash, ipfsCid, arweaveTxn, metadataURI, txHash: receipt.hash },
      null,
      2
    )
  );
}

main().catch((err) => {
  console.error("❌ Error:", err.message ?? err);
  process.exit(1);
});

File: scripts/upload_and_mint.py
#!/usr/bin/env python3
"""
Upload AI model weights to IPFS/Arweave and mint ERC-721 AI token in one flow.

Usage:
  # Upload to IPFS
  python scripts/upload_and_mint.py --weights model.onnx --name "My Model" --storage ipfs

  # Upload to Arweave  
  python scripts/upload_and_mint.py --weights model.onnx --name "My Model" --storage arweave

  # Upload to both (recommended for redundancy)
  python scripts/upload_and_mint.py --weights model.onnx --name "My Model" --storage both

  # Dry run (just generate metadata, no upload)
  python scripts/upload_and_mint.py --weights model.onnx --name "My Model" --architecture "ResNet-50" --dataset-hash abc123 --dry-run

Requires:
  - ipfs-http-client (pip install ipfshttpclient)
  - arweave-python (pip install arweave)
  - web3 (pip install web3)
"""

import argparse
import hashlib
import json
import logging
import os
import sys
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Optional

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")
logger = logging.getLogger(__name__)


@dataclass
class ModelMetadata:
    """ERC-721 AI token metadata as specified in issue #4.

    tokenURI should point to JSON with:
    - model hash (SHA-256)
    - storage CID (IPFS) or txn ID (Arweave)
    - architecture description
    - training dataset hash
    """
    name: str
    description: str
    model_hash_sha256: str
    storage_cid: str  # IPFS CID or Arweave TX ID
    storage_type: str  # "ipfs" or "arweave"
    architecture: str
    training_dataset_hash: str
    version: str = "1.0.0"

    def to_token_uri_json(self) -> str:
        """Generate tokenURI JSON content."""
        return json.dumps(asdict(self), indent=2)


def compute_sha256(file_path: str) -> str:
    """Compute SHA-256 hash of a file."""
    sha256 = hashlib.sha256()
    with open(file_path, "rb") as f:
        for chunk in iter(lambda: f.read(8192), b""):
            sha256.update(chunk)
    return sha256.hexdigest()


# ── IPFS Upload ────────────────────────────────────────────────────────────────

def upload_to_ipfs(file_path: str) -> str:
    """Upload file to IPFS and return the CID.

    Requires a running IPFS node or pinning service.
    """
    try:
        import ipfshttpclient
    except ImportError:
        logger.error("ipfshttpclient not installed. Run: pip install ipfshttpclient")
        sys.exit(1)

    logger.info(f"Uploading {file_path} to IPFS...")

    try:
        with ipfshttpclient.connect() as client:
            result = client.add(file_path)
            cid = result["Hash"]
            logger.info(f"IPFS upload complete. CID: {cid}")
            return cid
    except Exception as e:
        logger.error(f"IPFS upload failed: {e}")
        logger.info("Make sure IPFS daemon is running: ipfs daemon")
        sys.exit(1)


# ── Arweave Upload ─────────────────────────────────────────────────────────────

def upload_to_arweave(file_path: str, wallet_path: Optional[str] = None) -> str:
    """Upload file to Arweave and return the transaction ID.

    Requires an Arweave wallet (keyfile JSON).
    """
    try:
        from arweave.arweave_lib import Wallet, Transaction
    except ImportError:
        logger.error("arweave not installed. Run: pip install arweave")
        sys.exit(1)

    if not wallet_path:
        wallet_path = os.environ.get("ARWEAVE_WALLET_PATH")
    if not wallet_path:
        logger.error("Arweave wallet path required. Set ARWEAVE_WALLET_PATH or pass --wallet")
        sys.exit(1)

    logger.info(f"Uploading {file_path} to Arweave...")

    try:
        wallet = Wallet(wallet_path)
        with open(file_path, "rb") as f:
            data = f.read()

        tx = Transaction(wallet, data=data)
        tx.add_tag("Content-Type", "application/octet-stream")
        tx.add_tag("App-Name", "ERC721-AI-Weights")
        tx.sign()
        tx.send()

        logger.info(f"Arweave upload complete. TX: {tx.id}")
        return tx.id
    except Exception as e:
        logger.error(f"Arweave upload failed: {e}")
        sys.exit(1)


# ── Metadata Upload ────────────────────────────────────────────────────────────

def upload_metadata_to_ipfs(metadata: ModelMetadata) -> str:
    """Upload metadata JSON to IPFS and return CID for tokenURI."""
    try:
        import ipfshttpclient
        import tempfile

        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            f.write(metadata.to_token_uri_json())
            meta_path = f.name

        with ipfshttpclient.connect() as client:
            result = client.add(meta_path)
            cid = result["Hash"]
            os.unlink(meta_path)
            return f"ipfs://{cid}"
    except Exception as e:
        logger.warning(f"Could not upload metadata to IPFS: {e}")
        # Return data URI as fallback
        import base64
        encoded = base64.b64encode(metadata.to_token_uri_json().encode()).decode()
        return f"data:application/json;base64,{encoded}"


# ── Main Flow ──────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description="Upload AI model weights and mint ERC-721 token")
    parser.add_argument("--weights", required=True, help="Path to model weights file")
    parser.add_argument("--name", required=True, help="Model name")
    parser.add_argument("--description", default="", help="Model description")
    parser.add_argument("--architecture", required=True, help="Architecture description (e.g., 'ResNet-50, PyTorch')")
    parser.add_argument("--dataset-hash", required=True, help="SHA-256 hash of training dataset")
    parser.add_argument("--storage", choices=["ipfs", "arweave", "both"], default="ipfs",
                        help="Storage backend (default: ipfs)")
    parser.add_argument("--wallet", help="Arweave wallet keyfile path")
    parser.add_argument("--dry-run", action="store_true", help="Skip actual upload, just generate metadata")

    args = parser.parse_args()

    # Step 1: Compute model hash
    logger.info("Computing model SHA-256 hash...")
    model_hash = compute_sha256(args.weights)
    logger.info(f"Model hash: {model_hash}")

    # Step 2: Upload weights
    storage_cid = ""
    storage_type = args.storage

    if args.dry_run:
        storage_cid = "QmDRUMTQcVYUFPGn466uEtiGC8jU7bjhMiR7Y3iDSqTTNn"
        logger.info(f"[DRY RUN] Would upload to {storage_type}")
    elif args.storage in ("ipfs", "both"):
        storage_cid = upload_to_ipfs(args.weights)
        storage_type = "ipfs"
    elif args.storage == "arweave":
        storage_cid = upload_to_arweave(args.weights, args.wallet)
        storage_type = "arweave"

    if args.storage == "both" and not args.dry_run:
        # Also upload to Arweave for redundancy
        ar_tx = upload_to_arweave(args.weights, args.wallet)
        logger.info(f"Redundant copy on Arweave: {ar_tx}")

    # Step 3: Build metadata
    metadata = ModelMetadata(
        name=args.name,
        description=args.description,
        model_hash_sha256=model_hash,
        storage_cid=storage_cid,
        storage_type=storage_type,
        architecture=args.architecture,
        training_dataset_hash=args.dataset_hash,
    )

    logger.info("Generated metadata:")
    print(metadata.to_token_uri_json())

    # Step 4: Upload metadata to IPFS for tokenURI
    if not args.dry_run:
        token_uri = upload_metadata_to_ipfs(metadata)
        logger.info(f"tokenURI: {token_uri}")
    else:
        logger.info("[DRY RUN] Would upload metadata to IPFS for tokenURI")


if __name__ == "__main__":
    main()
